### DISCUSSION TITLE
Support MFA/2-step verification

### DISCUSSION THREAD

**Discussion by ptz0n** [ptz0n/homebridge-verisure#107] [Ideas]
> __Until this feature is implemented, we suggest that you disable MFA for your Verisure user. Please see [instructions](https://github.com/ptz0n/homebridge-verisure#multi-factor-authentication).__

```bash
Error: Cookie missing from response: HTTP 200
```

## MFA/2FA flow

1. `POST /auth/login` using basic auth, as we always done to get step up token
2. `POST /auth/mfa` using step up token cookie `vs-stepup`
3. `POST /auth/mfa/validate` using the same step up token cookie and pass the SMS code as body `{"token":"<code>"}` to get auth cookie
4. `POST /auth/trust` using the auth cookie to get long lived trust cookie `vs-trust-<redacted>`

__Suggested solution:__

In order to retrieve the `vid` cookie, we could provide an interactive cli tool:

```bash
$ ./cli.js 
verisure: email:  foo@bar.com
verisure: password:

 One-time code sent. 

verisure: code:  FAKE12

 Token for config: "vid=<redacted>"
```

---

🙌  Please contribute by sharing your ideas and comments.

**Comment by ptz0n** [comment]
__Confirmed:__ We should persist and use the `refreshToken`.

```bash
curl 'https://m-api02.verisure.com/auth/token' -X 'POST' \
  -H 'Content-Length: 0' \
  -H 'Cookie: vs-refresh=<refreshToken>'
```

```json
{
  "accessToken": "<redacted>",
  "accessTokenMaxAgeSeconds": 900,
  "refreshToken": "<redacted>",
  "refreshTokenMaxAgeSeconds": 63068527
}
```

**Comment by ptz0n** [comment]
And we can use the `accessToken` to perform requests.

```bash
curl 'https://m-api02.verisure.com/graphql' \
  -H 'Cookie: vid=<legacyToken>; vs-access=<accessToken>' \
  --data-raw $'{
    "operationName": "GetInstallation",
    "variables": {
      "giid": "<redacted>"
    },
    "query": "query GetInstallation($giid: String!) { installation(giid: $giid) { alias } }"
  }'
```

```json
{
  "data": {
    "installation": {
      "alias": "Strandvägen"
    }
  }
}
```

**Comment by palletodap** [comment]
Be aware that the function for disabling verification on Versiures website seems to only work in English language.

Every time i tried at first in Danish, it just reloaded the page. After changing to English it worked perfectly.

**Comment by ptz0n** [comment]
Progress. I've now tested the MFA flow and can confirm that we still only need the `vid` cookie to make authenticated API requests. This one, according to the set cookie header, will expire after one (1) year. It's unknown if it could expire as a result of any other event.

- __`vs-stepup`__ cookie is required for validating the one time SMS code. Expires after 20 minutes.
- __`vs-access`__ cookie is required to retrieve a `vs-trust` cookie, which could be helpful if we want to rotate the `vid` cookie using only username & password. Expires after 15 minutes.
- __`vs-refresh`__ cookie can be used to retrieve a new token using the endpoint `/auth/token`. Expires after 2 years.
- __`vs-trust-…`__ cookie can be used when authenticating with username and password to bypass MFA. Expires after ~68 years.

**Comment by ptz0n** [comment]
Just released `v1.14.0` with support for MFA. Happy Wednesday 🎉

**Reply by redjoto** [reply]
Thumbs up! 👍🏼

**Reply by hoefs** [reply]
Works great!!! Thanks!!!

**Reply by pierre-jean42** [reply]
Hello, I just installed your plugin and am very excited to try it, unfortunately I have not figured out how to complete the installation. Presumably because of the 2FA on my account. I tried running npx homebridge-verisure to get the code, but instead I received Error: Request failed with status code 401

Do you happen to know what I should do? I'm running homebridge from an M1 iMac. Thanks eitherway!
